### ISSUE TITLE
Performance issues when having 100 items in ShoppingCart

### ISSUE THREAD

**Issue by brakim** [nopSolutions/nopCommerce#3363]
Hi, i'm currently building a B2B shop and some customer place pretty large order (many products), in some rare cases over 200 products but pretty much all of them is between 50-100 products.

After some debugging the whole /Cart method i found 2 big performance impacts in the ShoppingCartModelFactory, and i guess both gets slow because of lazy loading.

Line 372:
`cartItemModel.DisableRemoval = cart.Any(item => item.Product.RequireOtherProducts && _productService.ParseRequiredProductIds(item.Product).Contains(sci.ProductId));`

Takes 1086 ms

Line 454:
`var itemWarnings = _shoppingCartService.GetShoppingCartItemWarnings(
                _workContext.CurrentCustomer,
                sci.ShoppingCartType,
                sci.Product,
                sci.StoreId,
                sci.AttributesXml,
                sci.CustomerEnteredPrice,
                sci.RentalStartDateUtc,
                sci.RentalEndDateUtc,
                sci.Quantity,
                false,
                sci.Id);`

Takes 1132 ms

Both of these are run per product, so a total of around 200+ seconds to load the /Cart page with 100 products.

nopCommerce version: 4.10

Steps to reproduce the problem:
1. Clean installation
2. Create products 
3. Place them all in the shopping cart
4. Visit /cart

**Comment by aschiebold** [comment]
It's because most of the transactions in nopcommerce make hundreds of SQL queries per transaction. A simple load of the shopping cart has an insanely huge call stack and number of queries. Here's a post I opened on the forums:

https://www.nopcommerce.com/boards/t/55364/i-installed-appdynamics-on-my-site.aspx

**Comment by brakim** [comment]
@aschiebold how many products was in the cart during your trace?

**Comment by aschiebold** [comment]
@brakim IIRC it was just one item. Worst case it was maybe 3 items.

**Comment by brakim** [comment]
@aschiebold its funny that it calls **GetShoppingCartSubTotal** 12 times for one item.

**Comment by Leftyx** [comment]
@brakim I have the same problem with categories and manufacturers.
When I first load the home page I can see 600 queries running. I have never seen anything like that before ... but that's the way it is.
If you have localized properties it would run one query for each localized field.

One easy fix to reduce the amount of queries would be to comment some of the properties you are not interested to be localized:

```
.Select(category =>
{
    var catModel = new CategoryModel
    {
	Id = category.Id,
	Name = _localizationService.GetLocalized(category, x => x.Name),
	Description = _localizationService.GetLocalized(category, x => x.Description),
	//MetaKeywords = _localizationService.GetLocalized(category, x => x.MetaKeywords),
	//MetaDescription = _localizationService.GetLocalized(category, x => x.MetaDescription),
	//MetaTitle = _localizationService.GetLocalized(category, x => x.MetaTitle),
	SeName = _urlRecordService.GetSeName(category),
    };
    
    ...
    
})
.ToList();
```
There's not real view model in NopCommerce. Most of the entities are fully loaded and related entities are fetch one by one (SELECT + 1).
I've tried to fix some of these bits but at some point I banged my head against a wall.
There are too many useless layers. They've used a repository patter with Entity Framework which is already a repository.
Queries are in this Factory layer when they should be sitting in the service layer.
At the moment I've asked my client to add RAM and processors but at some point I might need to look somewhere else.

**Comment by hellcoderGIT** [comment]
I have added all 3 performance optimizations of AndreiMaz and the site is still unusable with more than ~20 items in the cart (takes about 2 minutes to load the shopping cart) :/
When I checked the amount of queries for one shopping cart load with about 20 items it was more than 500, in an azure cloud setup that just doesn't work very well.

tried upgrading the database to S2 but it only made it marginally better. 

I think some serious refactoring needs to be done to fix this, aggressive caching of everything that is already in the shopping cart maybe?

 https://www.advmotostickers.com/en/

**Comment by brakim** [comment]
@hellcoderGIT the performance is also related to this #3566.

Me and @kwestground have been picking NopCommerce services to and factories to pieces and found multiple leeks where some things are not cached and when they are cached the cpu cycles is just through the roof. Specially when you combine many tier prices linked to many different customer roles, and then add these products to the cart.

**Comment by hjvanreenenPN** [comment]
In your own site on https://demo.nopcommerce.com/ the shopping cart is still very slow when adding more than 20 products.

Are the latest commits merged into the demo store?

Steps to reproduce:
- go to https://demo.nopcommerce.com/
- add 20 products to the shopping cart
- open the shopping cart 
Result:
very slow

**Comment by AndreiMaz** [comment]
@hjvanreenenPN it's still powered by the previous version. I've talked about the new upcoming one - 4.20

**Comment by hjvanreenenPN** [comment]
When will 4.20 be released? (We have more stability and performance related issues on the webshop of our client)

**Comment by AndreiMaz** [comment]
Please see https://www.nopcommerce.com/boards/t/55550/nopcommerce-420.aspx?p=3#227459

**Comment by brakim** [comment]
@AndreiMaz i tried upgrading the site to 4.20, the performance is slightly better.
Our test cases is a customer with 100 products in the cart, the customer is assigned to a role that is connected to some discount and also tier prices.

Each product have about 15 different tier prices, we use tier price in combination with customer role to mimic a customer unique price list.

With this setup the shopping cart takes around ~120 seconds to load.
If i delete all discounts and all tier prices the loading time goes down to 340 ms which is good.

If you want i can provide our test database.

**Comment by iamexey** [comment]
@AndreiMaz this issue is very real for us aswell, ur commited fixes for 4.2 made it go from 89 to 88 seconds with a big cart. We just got straight prices no discountcodes or anything of the sort.

Anyone with some other idea?
